### PR TITLE
Raise errors when dsl is misused

### LINKED ISSUE #394: Running Individual Specs
Hey guys, how can I run individual specs with Cedar and not the whole test suite?

### PR DESCRIPTION
**Pull request by tjarratt**
Fixes #394.

The first commit is the brunt of this feature work. DSL misuse encompasses two main failure modes:

* someone calls `it()`, `describe()`, or `context()` inside of an `it()` block
  * this would never work because the newly-added tests are never run
* someone tries to call any of the above functions outside of a test
  * this would never work because there is no spec yet to add them to until we build the tree

In writing this feature I tried to avoid adding much more scope to Cedar's public interface, but found it necessary to expose `CDRdisableSpecValidation()` and `CDRenableSpecValidation()` functions so that we could keep our existing tests that create cedar examples by calling `it()` inside of a test. I also thought it desirable to allow someone to disable this feature, if they were using Cedar in some interesting way.

Then, once I was totally done with this feature, writing a happy commit message and running all of the test suite ... I discovered a completely unrelated regression in the Type Utilities part of Cedar. The commit message for that is pretty verbose -- I recommend reading it if you want to know when a valid pointer sometimes can be misconstrued to be a NULL pointer. It was quite a find.

### CODE REVIEW

**Comment by modocache** [comment]
Cool! Quick added something similar in https://github.com/Quick/Quick/pull/448, and I think that reduced the number of issues being reported on the project.

**Comment by tjarratt** [comment]
@idoru I'd be curious to know if you have any feedback or thoughts around this PR. I've attempted to be a bit more conservative than I normally would when refactoring and testing with mocks, but felt like some of this was necessary.

I've been feeling some pain around the maintainability of `CDRfunctions.{h|m}` and feel like starting to extract some of this to Obj-C when possible would make testing parts of the implementation a tad easier, although I suspect you could teach me a thing or two about testing Cedar with Cedar 😉

File: Source/CDRSpec.m [hunk 1]
```diff
@@ -1,15 +1,46 @@
...
-static void(^placeholderPendingTestBlock)() = ^{ it(@"is pending", PENDING); };
+#pragma mark - Spec Validation
+
+// runs validation when it(), context(), describe() are invoked
+// generally this is a useful feature, but you may disable it if you
+// care to do metaprogramming with Cedar specs
+// (or other usecases we couldn't imagine).
+BOOL CDR_validateSpecs = YES;
 
+void CDRdisableSpecValidation() {
```
Comment (idoru) [inline Source/CDRSpec.m]: Capitalization on `CDRdisable...`?

File: Source/CDRSpec.m [hunk 2]
```diff
@@ -1,15 +1,46 @@
...
+    CDR_validateSpecs = NO;
+}
+
+void CDRenableSpecValidation() {
```
Comment (idoru) [inline Source/CDRSpec.m]: Ditto Capitalization?

File: Source/CDRRunState.h [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#import <Foundation/Foundation.h>
+
+typedef NS_ENUM(NSInteger, CedarRunState) {
+    CedarRunStateNotYetStarted = 0,
+    CedarRunStateBuildingTree  = 1,
```
Comment (idoru) [inline Source/CDRRunState.h]: Not sure `...Tree` on this is useful or relevant information for this type.

File: Source/CDRSpec.m [hunk 4]
```diff
@@ -18,43 +49,96 @@ void afterEach(CDRSpecBlock block) {
...
+            [[NSException exceptionWithName:NSInternalInconsistencyException
+                                     reason:reason
+                                   userInfo:nil] raise];
+            break;
         }
-        CDR_currentSpec.currentGroup = parentGroup;
-    } else {
-        group = describe(text, placeholderPendingTestBlock);
+        case CedarRunStateFinished:
+            // someone ... REALLY done goofed
```
Comment (idoru) [inline Source/CDRSpec.m]: Should there be an error raised here, given this comment? Is it even possible to trigger this?

File: Source/CDRStateTracker.m [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#import "CDRStateTracker.h"
+#import "CDRRunState.h"
+
+@implementation CDRStateTracker
+
+- (void)didStartBuildingTestTree {
```
Comment (idoru) [inline Source/CDRStateTracker.m]: Similar to my comments about the enum; not sure the fact that we're building a tree structure is useful or relevant.

**Comment by idoru** [comment]
@tjarratt thanks for doing this work! I had a few naming nit-picks which I've commented on c1f33d2.

Regarding 2ddf819... How interesting. It's interesting that we chose to compare with '\0' to check for a NULL pointer. Perhaps we should compare with (void *)0?

Checkout §6.3.2.3 point 3 in particular: http://www.open-std.org/JTC1/SC22/WG14/www/docs/n1124.pdf

File: Source/CDRRunState.h [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#import <Foundation/Foundation.h>
```
Comment (briancroom) [inline Source/CDRRunState.h]: This file should go into the appraise `Headers` directory to make sure Cocoapods finds it.

File: Source/CDRRunState.h (hunk 3 above)
Comment (tjarratt) [inline Source/CDRRunState.h]: I mostly wanted to clarify since "building" could be overloaded in the context of Xcode, building projects and targets and whatnot.

Looking at it again... I wonder if there's a better phrase that would indicate users what is actually happening here. Perhaps `CedarRunStatePreparingToRunTests` ?

File: Source/CDRStateTracker.m (hunk 5 above)
Comment (tjarratt) [inline Source/CDRStateTracker.m]: That's a good point, although this is an internal implementation that isn't intended to be exposed. This should be a project-level header with no visibility at all.

I think this should stay consistent with however we end up naming the enum, however, so I'm happy to improve this too.

File: Source/CDRSpec.m (hunk 4 above)
Comment (tjarratt) [inline Source/CDRSpec.m]: Absolutely, although it's pretty tough to trigger this in a realistic test, since you'd need to call `it()` and assert that an exception is raised sometime after `main()` and before Cedar starts to run.

**Comment by tjarratt** [comment]
Thanks for all the feedback, @idoru. Brian commented on an outdated diff that comparison to `NULL` would be better than `'\0'`. The definition for `NULL` is a bit hard to follow, but it seems like it will defined as `((void*)0)` when C++ is not defined, and `0` otherwise. 

`NULL` seems like a perfectly cromulent constant to here use, but I'm open to switching it to an explicit `(void *)0` too.
